File: Medium/314.py
from collections import deque, defaultdict
from typing import Optional, List


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def verticalOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        output = []

        if root is None:
            return output
        
        # list：当 key 不存在的时候，自动创建一个默认 value。
        # column_table = defaultdict(list)
        column_table = {}
        queue = deque([(root, 0)])

        while queue:
            node, column = queue.pop()

            if node is not None:
                if column not in column_table:
                    column_table[column] = []
                column_table[column].append(node.val)
                queue.appendleft((node.left, column - 1))
                queue.appendleft((node.right, column + 1))

        sorted_keys = sorted(column_table.keys())

        for k in sorted_keys:
            output.append(column_table[k])

        return output

File: Medium/314.cs
using System;
using System.Collections.Generic;
using System.Linq;

// Definition for a binary tree node.
// public class TreeNode {
//     public int val;
//     public TreeNode left;
//     public TreeNode right;
//     public TreeNode(int val = 0, TreeNode left = null, TreeNode right = null) {
//         this.val = val;
//         this.left = left;
//         this.right = right;
//     }
// }

public class Solution {
  public IList<IList<int>> VerticalOrder(TreeNode root) {
    IList<IList<int>> output = new List<IList<int>>();

    if (root == null) {
      return output;
    }

    Dictionary<int, List<int>> columnTable = new Dictionary<int, List<int>>();
    Queue<(TreeNode node, int column)> queue = new Queue<(TreeNode node, int column)>();

    queue.Enqueue((root, 0));

    while (queue.Count != 0) {
      var p = queue.Dequeue();
      TreeNode node = p.node;
      int column = p.column;

      if (!columnTable.ContainsKey(column)) {
        columnTable[column] = new List<int>();
      }

      columnTable[column].Add(node.val);

      if (node.left != null) {
        queue.Enqueue((node.left, column - 1));
      }

      if (node.right != null) {
        queue.Enqueue((node.right, column + 1));
      }
    }

    List<int> sortedKeys = columnTable.Keys.OrderBy(k => k).ToList();

    foreach (int k in sortedKeys) {
      output.Add(columnTable[k]);
    }

    return output;
  }
}